core/src/main/java/smile/classification/DecisionTree.java

```diff
@@ -997,6 +985,8 @@ public DecisionTree(Attribute[] attributes, double[][] x, int[] y, int maxNodes,
             node.split(nextSplits); // Split the parent node into two children nodes
         }
 
+        pruneRedundantLeaves(root);
```

Comment:
I think post pruning should be used for best-first splitting at https://github.com/haifengl/smile/pull/449/files#diff-aa1c5dce2c959f46efdcb90067706154R1025 when https://github.com/haifengl/smile/pull/454 is merged.

https://github.com/haifengl/smile/pull/454 does pruning while building a tree in depth-first and thus the maximum memory consumption is reduced.

And, post-pruning could be generalized to do the simple pruning or to do reduced error (say more aggressive) pruning https://github.com/haifengl/smile/pull/360 by a configuration.

Reply:
You're right; we can do pruning while building the tree, and then also do a post-pass to re-merge any redundant splits, which can still happen in best-first and depth-first splitting.

Reduced error pruning is worth exploring later.